### PR TITLE
feat(guards): more control over undefined and null checks

### PR DESCRIPTION
**Pull request by eranhirsch**
When working with stricter typescript null and undefined are not the same, and sometimes you'd want to filter a type `{} | null | undefined` so that you still retain the null or undefined values.

To prevent any backwards compatibility issues I created a new variant for isDefined for the explicit case, and added an `isNonNull` for nulls.

I also split the guard tests so that they are colocated with the files following the split done recently.
---
- [x] Typedoc added for new methods and updated for changed
- [x] Tests added for new methods and updated for changed
- [x] New methods added to `src/index.ts`

### CODE REVIEW

File: src/isDefined.ts [hunk 1]
```diff
@@ -13,3 +13,9 @@
 export function isDefined<T>(data: T): data is NonNullable<T> {
   return typeof data !== 'undefined' && data !== null;
 }
+
+export namespace isDefined {
+  export function strict<T>(data: T | undefined): data is T {
```
Comment (TkDodo) [inline src/isDefined.ts:18]: we'd also need docs here

File: src/isNonNull.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export function isNonNull<T>(data: T | null): data is T {
+  return data !== null;
+}
```
[the remark this answers is not in the document]
Comment (eranhirsch) [reply src/isNonNull.ts:16]: oops, i had another commit i never pushed with the docs and index changes 😨

File: src/isDefined.ts (hunk 1 above)
Comment (eranhirsch) [reply src/isDefined.ts:23]: do we document variants? I couldn't find an example where we do, the documentation usually resides on the main object

Comment (TkDodo) [reply src/isDefined.ts:23]: ah you're right, it's enough if we have an example 👍

**Review by TkDodo — APPROVED** [review]

**Pull request merged**